BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c

```diff
@@ -352,3 +352,55 @@ GetFeatureCfg (
   return GetLoaderGlobalDataPointer()->LdrFeatures;
 }
 
+
+/**
+  Get device address
+
+  @param[in]  DeviceType         The device type, refer OS_BOOT_MEDIUM_TYPE.
+  @param[in]  DeviceInstance     The device instance number starting from 0.
+  @param[in]  PciMmioAddr        Only applicable to PCI device. TRUE will return
+                                 PCI MMIO address, else return PCI BDF.
+
+  @retval     Device address for a given device instance, return 0 if the device
+              could not be found from device table.
+**/
+UINT32
+EFIAPI
+GetDeviceAddr (
+  IN  UINT8          DeviceType,
+  IN  UINT8          DeviceInstance,
+  IN  BOOLEAN        PciMmioAddr
+  )
+{
+  SBL_DEVICE_TABLE   *DeviceTable;
+  SBL_DEVICE         *Device;
+  UINT32             DeviceBase;
+  UINT32             Index;
+
+  DeviceBase = 0;
+  DeviceTable = ((LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer())->DeviceTable;
+  for (Index = 0; Index < DeviceTable->DeviceNumber; Index++) {
+    Device = &DeviceTable->Device[Index];
+    if ((Device->Type == DeviceType) && (Device->Instance == DeviceInstance)){
+      break;
+    }
+  }
+
+  if (DeviceTable->DeviceNumber == Index) {
+    // return 0 if the device is not found in device table.
+    return DeviceBase;
+  }
+
+  if ((Device->Dev.PciDev.IsMmioDevice == 0) && PciMmioAddr) {
+    DeviceBase = (UINTN)MM_PCI_ADDRESS (
+                          Device->Dev.PciDev.PciBusNumber,
+                          Device->Dev.PciDev.PciDeviceNumber,
+                          Device->Dev.PciDev.PciFunctionNumber,
+                          0);
+  } else {
+    DeviceBase = Device->Dev.DevAddr;
+  }
+
+  return DeviceBase;

```

Comment:
can we always return Device->Dev.DevAddr ?

Reply:
OK. will add another Macro to convert to PCI base address.